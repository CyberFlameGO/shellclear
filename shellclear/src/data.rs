use crate::shell::Shell;
use serde_derive::Deserialize;

pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
}

#[derive(Debug)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
}
