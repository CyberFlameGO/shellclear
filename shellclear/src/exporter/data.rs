use crate::data::FindingSensitiveCommands;
use crate::shell::Shell;
use anyhow::Result;
use chrono::{DateTime, Local, NaiveDateTime, TimeZone};
use lazy_static::lazy_static;
use regex::Regex;

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";

pub trait Exporter {
    fn sensitive_data(&self, findings: &[&FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

// todo:: move under export trait
pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_TIME.captures(&finding.data) {
                if let Some(timestemp) = c.get(1) {
                    let date_time: DateTime<Local> = Local.from_utc_datetime(
                        &NaiveDateTime::from_timestamp(timestemp.as_str().parse::<i64>()?, 0),
                    );
                    return Ok(format!("{}", date_time.format(DATE_TIME_FORMAT)));
                }
            };
            Ok(finding.data.clone())
        }
        _ => Ok("".to_string()),
    }
}
