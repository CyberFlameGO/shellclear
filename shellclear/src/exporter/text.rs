use crate::data::FindingSensitiveCommands;
use crate::exporter::data::{chunk, extract_time, Exporter};
use anyhow::Result;
use console::style;

#[derive(Default)]
pub struct Text {}

impl Exporter for Text {
    fn sensitive_data(&self, findings: &[&FindingSensitiveCommands]) -> Result<()> {
        let mut count = 0;
        for f in findings.iter() {
            count += 1;
            let finding_names = f
                .sensitive_findings
                .iter()
                .map(|f| f.name.clone())
                .collect::<Vec<_>>()
                .join(",");

            let title = format!(
                "{}. {} {}",
                count,
                finding_names,
                extract_time(f).unwrap_or_else(|_| "".to_string())
            );
            println!("{}", style(title).bold());
            println!("{}", chunk(&f.command, 150));
        }
        Ok(())
    }
}
