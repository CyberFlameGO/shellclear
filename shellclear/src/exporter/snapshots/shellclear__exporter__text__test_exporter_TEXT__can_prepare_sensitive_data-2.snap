---
source: shellclear/src/exporter/text.rs
expression: "str::from_utf8(&out).unwrap().replace(\"\\r\\n\", \"\\n\")"
---
"\u{1b}[1m1. test name,test name2 2022-06-13 11:55:59\u{1b}[0m\ntest command\n"
